# Compléments: Opérations 'bitwise'

Les compléments ci-dessous expliquent des fonctions évoluées sur les entiers; les débutants en programmation peuvent sans souci sauter cette partie en cas de difficultés.

### Opérations logiques : ET &, OU | et OU exclusif ˆ

Il est possible aussi de faire des opérations "bit-à-bit" sur les nombres entiers. 
Le plus simple est de penser à l'écriture du nombre en base 2

Considérons par exemple deux entiers constants dans cet exercice:

In [1]:
x49 = 49
y81 = 81

Ce qui nous donne comme décomposition binaire 

$\begin{array}{rcccccc}
x49&=&49&=&32 + 16 + 1 &\rightarrow&(0,1,1,0,0,0,1) \\
y81&=&81&=&64 + 16 + 1 &\rightarrow&(1,0,1,0,0,0,1)
\end{array}
$

##### Et logique : opérateur &

L'opération logique '&' va faire un 'et' 
logique bit à bit entre les opérandes, ainsi

In [2]:
x49 & y81

17

Et en effet
$\begin{array}{rcl}
x49 & \rightarrow & (0,1,1,0,0,0,1) \\
y81 & \rightarrow & (1,0,1,0,0,0,1) \\
x49 \& y81 & \rightarrow & (0,0,1,0,0,0,1) \rightarrow 17
\end{array}$


##### Ou logique : opérateur |

De même, l'opération logique '|' fait simplement un 'ou' logique, comme ceci

In [3]:
x49 | y81

113

On s'y retrouve parce que
$\begin{array}{rcl}
x49 & \rightarrow & (0,1,1,0,0,0,1) \\
y81 & \rightarrow & (1,0,1,0,0,0,1) \\
x49 | y81 & \rightarrow & (1,1,1,0,0,0,1) \rightarrow 64+32+16+1 \rightarrow 113
\end{array}$


##### Ou exclusif : opérateur ^

Enfin on peut également faire la même opération à base de 'ou exclusif' avec l'opérateur 'ˆ'

In [4]:
x49 ^ y81

96

### Décalages

Un décalage 'à gauche' de, par exemple, 4 positions, revient à décaler tout le champ de bits de 4 cases à gauche (les 4 nouveaux bits insérés sont toujours des 0); c'est donc équivalent à une multiplication par $2ˆ4=16$

In [5]:
x49 << 4

784

$\begin{array}{rcl}
x49 & \rightarrow & (0,1,1,0,0,0,1) \\
x49 << 4 & \rightarrow & (0,1,1,0,0,0,1,0,0,0,0) \rightarrow 512+256+16 \rightarrow 784 
\end{array}$


De la même façon le décalage à droite de $n$ revient à une
division par $2ˆn$ (et plus exactement le quotient de la division)

In [6]:
x49 >> 4

3

$\begin{array}{rcl}
x49 & \rightarrow & (0,1,1,0,0,0,1) \\
x49 >> 4 & \rightarrow & (0,0,0,0,0,1,1) \rightarrow 2+1 \rightarrow 3
\end{array}$

### Une astuce

On peut utiliser la fonction 'built-in' bin 
pour calculer la représentation binaire d'un entier; attention la valeur de retour est un 'string'

In [7]:
bin(x49)

'0b110001'

Dans l'autre sens, on peut aussi entrer un entier directement en base 2 comme ceci; ici comme on le voit $x49bis$ est bien un entier

In [8]:
x49bis = 0b110001
x49bis == x49

True

### Pour en savoir plus

https://docs.python.org/2/library/stdtypes.html#bitwise-operations-on-integer-types